feat: add C implementation for `math/base/special/betainc`

Neerajpathak07 opened the pull request:
Progresses #649 

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for `math/base/special/betainc`

## Related Issues

> Does this pull request have any related issues?

This pull request:

- progresses #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: add-betainc

Neerajpathak07 commented:
This PR will need the C implementation for `math/base/special/kernel-betainc` to be merged to develop.

Neerajpathak07 commented:
Reference:-
![image](https://github.com/user-attachments/assets/8a854034-ffa6-4de7-ad4b-c8ab85896952)

rautelaKamal commented:
Hey @Neerajpathak07, was reviewing this since it depends on kernel-betainc which I've been following closely. Found a few things in src/main.c that might block compilation:

Line 52 uses double out = [ 0.0, 0.0 ] which is JS syntax - in C this would need to be separate variable declarations like double out; double deriv;

Lines 53-54 redeclare regularized and upper, which are already function parameters - compiler will error there.

Line 55 calls stdlib_base_kernelBetainc but the actual C function in #10279 is stdlib_base_kernel_betainc (snake_case), and it expects pointer arguments (&out, &deriv) rather than array and stride.

The header include on line 27 should be kernel_betainc.h instead of kernelbetainc.h.

Also minor - the @returns JSDoc says "natural logarithm of beta function" but the function returns the incomplete beta value, worth updating to avoid confusion.

Hope this helps move things forward!

rautelaKamal commented:
Hey @Neerajpathak07, was reviewing this since it depends on kernel-betainc which I have been following closely. Found a few things in src/main.c that might block compilation:

Line 52 uses double out = [ 0.0, 0.0 ] which is JS syntax, in C this would need to be separate variable declarations like double out; double deriv;

Lines 53-54 redeclare regularized and upper, which are already function parameters, compiler will error there.

Line 55 calls stdlib_base_kernelBetainc but the actual C function in #10279 is stdlib_base_kernel_betainc (snake_case), and it expects pointer arguments (&out, &deriv) rather than array and stride.

The header include on line 27 should be kernel_betainc.h instead of kernelbetainc.h.

Also minor, the @returns JSDoc says "natural logarithm of beta function" but the function returns the incomplete beta value, worth updating to avoid confusion.

Hope this helps move things forward!

rautelaKamal commented:
Hey @Neerajpathak07, was reviewing this since it depends on kernel-betainc which I have been following closely. Found a few things in src/main.c that might block compilation:

Line 52 uses double out = [ 0.0, 0.0 ] which is JS syntax, in C this would need to be separate variable declarations like double out; double deriv;

Lines 53-54 redeclare regularized and upper, which are already function parameters, compiler will error there.

Line 55 calls stdlib_base_kernelBetainc but the actual C function in #10279 is stdlib_base_kernel_betainc (snake_case), and it expects pointer arguments (&out, &deriv) rather than array and stride.

The header include on line 27 should be kernel_betainc.h instead of kernelbetainc.h.

Also minor, the @returns JSDoc says "natural logarithm of beta function" but the function returns the incomplete beta value, worth updating to avoid confusion.

Hope this helps move things forward!

rautelaKamal commented:
Hey @Neerajpathak07, was reviewing this since it depends on kernel-betainc which I have been following closely. Found a few things in src/main.c that might block compilation:

Line 52 uses double out = [ 0.0, 0.0 ] which is JS syntax, in C this would need to be separate variable declarations like double out; double deriv;

Lines 53-54 redeclare regularized and upper, which are already function parameters, compiler will error there.

Line 55 calls stdlib_base_kernelBetainc but the actual C function in #10279 is stdlib_base_kernel_betainc (snake_case), and it expects pointer arguments (&out, &deriv) rather than array and stride.

The header include on line 27 should be kernel_betainc.h instead of kernelbetainc.h.

Also minor, the @returns JSDoc says "natural logarithm of beta function" but the function returns the incomplete beta value, worth updating to avoid confusion.

Hope this helps move things forward!

rautelaKamal commented:
Hey @Neerajpathak07, was reviewing this since it depends on kernel-betainc which I have been following closely. Found a few things in src/main.c that might block compilation:

Line 52 uses double out = [ 0.0, 0.0 ] which is JS syntax, in C this would need to be separate variable declarations like double out; double deriv;

Lines 53-54 redeclare regularized and upper, which are already function parameters, compiler will error there.

Line 55 calls stdlib_base_kernelBetainc but the actual C function in #10279 is stdlib_base_kernel_betainc (snake_case), and it expects pointer arguments (&out, &deriv) rather than array and stride.

The header include on line 27 should be kernel_betainc.h instead of kernelbetainc.h.

Also minor, the @returns JSDoc says "natural logarithm of beta function" but the function returns the incomplete beta value, worth updating to avoid confusion.

Hope this helps move things forward!

Neerajpathak07 commented:
@rautelaKamal why do we have multiple duplicate comments here correct me if I am mistaken here but are you using any LLM's to perform code reviews?

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/betainc/src/main.c:
```diff
@@ -0,0 +1,57 @@
+*/
+
+#include "stdlib/math/base/special/betainc.h"
+#include "stdlib/math/base/special/kernelbetainc.h"
```
```suggestion
#include "stdlib/math/base/special/kernel_betainc.h"
```
Awaiting the necessary C implementation for the dependent package.

Neerajpathak07 replied:
Resolved as per #10279

0PrashantYadav0 commented:
Hi @Neerajpathak07 , since PR #10279 is merged now I think you can move ahead with implementation of `math/base/special/betainc`. Few of my PR depends on these.

Neerajpathak07 committed:
Merge branch 'stdlib-js:develop' into add-betainc

Neerajpathak07 committed:
chore: initial clean up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: passed
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 committed:
chore: sync C implementation with JS implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 committed:
chore: update C addon to return double instead of NULL

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 committed:
chore: remove const var for upper and regularized

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 committed:
Merge branch 'develop' into add-betainc

Planeshifter committed:
build: replace with standard include.gypi file

The current file contents are non-standard and caused an infinite loop: when node-gyp processed binding.gyp, it included include.gypi, which included include.gypi...

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Neerajpathak07 commented:
@Planeshifter accidently marked this open. I am on the changes and testing those to resolve the CI.

Neerajpathak07 committed:
refactor: native.js to add checks for regularized and upper

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 committed:
chore: update ULP values for tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: passed
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/betainc/benchmark/benchmark.js:
```diff
@@ -31,15 +31,19 @@ var betainc = require( './../lib' );
+	};
+	x = uniform( 100, 0.0, 1.0, opts );
+	a = uniform( 100, EPS, 1000.0, opts );
+	b = uniform( 100, EPS, 1000.0, opts );
```
Went ahead and updated the benchmarks and tests to follow the current code-convention.

Neerajpathak07 commented:
Since Lint failure is not related to changes in this PR. I'll go ahead and request a review.

Neerajpathak07 commented:
@Planeshifter Can we try and land this in the following week so as to unblock `beta` and `binomial` distributions.

Planeshifter committed:
style: fix missing final newlines and remove copy-paste artifacts

Planeshifter committed:
chore: remove incorrect license attribution and stray eslint comments

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/betainc/src/main.c:
```diff
@@ -0,0 +1,45 @@
+* @param a              function parameter
+* @param b              function parameter
+* @param regularized    boolean indicating if the function should evaluate the regularized or non-regularized incomplete beta function.
+* @param upper          boolean indicating if the function should return the upper tail of the incomplete beta function.
```
```suggestion
* @param regularized    boolean indicating if the function should evaluate the regularized or non-regularized incomplete beta function
* @param upper          boolean indicating if the function should return the upper tail of the incomplete beta function
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/betainc/src/main.c:
```diff
@@ -0,0 +1,45 @@
+	double derivative;
+
+	regularized = ( regularized == false ) ? false : true;
+	upper = ( upper == true ) ? true : false;
```
Doesn't make sense to keep because both parameter are boolean and have to be provided.

Planeshifter committed:
chore: apply review fixes for C implementation

- fix invalid arity in main.c doc example and use const bool params
- remove dead no-op boolean coercion in main.c
- use scalar out variable instead of 2-element array in main.c
- fix TAP label (c::cephes -> c::native) and benchmark default variant in C benchmark
- add explicit-variant benchmarks to benchmark.native.js matching labels
- remove stale wasm fields from manifest.json
- fix README C example annotation and signature, remove stray blank line
- reorder VARIABLES/FIXTURES sections and use t.strictEqual in test.native.js
- mirror main.js boolean coercion semantics in native.js
- add missing blank line after include guard in header

Planeshifter reviewed (approved):
Needed to do significant clean-up, but should be good to go now.

Note: We need to ensure the license comments are accurate; in this case, since this package is a thin wrapper around `kernel-betainc`, I removed the license comment and LICENSE file. But notice they were inconsistent with each other and incorrect.

Also, the EditorConfig lint failures were due to the changes in this PR, see https://github.com/stdlib-js/stdlib/pull/4037/changes/a77b0610f7becd1e6355c37f89dea6410fb927f8 which caused the checkmark to turn green.
